File: pset1/greedy.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

// prototypes
int coincount(int cents, int coin_value);
int coin_remain(int cents, int coin_value);

int main(void)
{
    // prompt user for change owed
    float input;
    do
    {
        printf("How much change is owed?:\n");
        input = GetFloat();
    }
    while (input < 0.0);
    
    // convert float into int in cents
    double clean_input = round(input * 100);
    int change = (int) clean_input;
    
    // analyze change, count coins, and pass remainder to same process with next coin
    int q_count = 0;
    int d_count = 0;
    int n_count = 0;
    int p_count = 0;
    if (change >= 25)
    {
        q_count = coincount(change, 25);
        change = coin_remain(change, 25);
    }
    if (change < 25 && change >=10)
    {
        d_count = coincount(change, 10);
        change = coin_remain(change, 10);
    }
    if (change < 10 && change >=5)
    {
        n_count = coincount(change, 5);
        change = coin_remain(change, 5);
    }
    if (change < 5)
    {
        p_count = change;
    }
    
    // calculate total number of coins and print
    int num_coin = q_count + d_count + n_count + p_count;
    printf("%i\n", num_coin);
}

// divides change by coin value and returns int of coins that can be divided into it
int coincount(int cents, int coin_value)
{
    return cents / coin_value;
}

// calculates and returns remainder of change leftover after coin value divided into change amount
int coin_remain(int cents, int coin_value)
{
    return cents % coin_value;
}

File: pset6-2017/greedy.py
import cs50

def main():
    # get positive float from user
    while True:
        print("How much change is owed ($)?: ")
        user_owed = cs50.get_float()
        if (user_owed >= 0):
            break

    # clean input and determine min number of coins
    cents = int(round(100 * user_owed))
    print(coin_count(cents))

def coin_count(cents):
    """Return minimum number of coins (quarters, dimes, and nickels) given 
    non-negative input in cents
    """
    if cents < 5:
        return cents
    else:
        if 5 <= cents < 10:
            coin = 5
        elif 10 <= cents < 25:
            coin = 10
        elif cents >= 25:
            coin = 25

        return (cents // coin) + coin_count(cents % coin)

if __name__ == "__main__":
    main()